client/finality-grandpa/src/until_imported.rs

```diff
@@ -47,30 +47,36 @@ use sp_finality_grandpa::AuthorityId;
 
 const LOG_PENDING_INTERVAL: Duration = Duration::from_secs(15);
 
-// something which will block until imported.
+/// Something that needs to be withheld until specific blocks are available.
+///
+/// For example a GRANDPA commit message which is not of any use without the corresponding block
+/// that it commits on.
 pub(crate) trait BlockUntilImported<Block: BlockT>: Sized {
-	// the type that is blocked on.
+	/// The type that is blocked on.
 	type Blocked;
 
-	/// new incoming item. For all internal items,
-	/// check if they require to be waited for.
-	/// if so, call the `Wait` closure.
-	/// if they are ready, call the `Ready` closure.
-	fn schedule_wait<S, Wait, Ready>(
+	/// Check if a new incoming item needs awaiting until a block(s) is imported.
+	fn needs_waiting<S: BlockStatusT<Block>>(
 		input: Self::Blocked,
 		status_check: &S,
-		wait: Wait,
-		ready: Ready,
-	) -> Result<(), Error> where
-		S: BlockStatusT<Block>,
-		Wait: FnMut(Block::Hash, NumberFor<Block>, Self),
-		Ready: FnMut(Self::Blocked);
+	) -> Result<DiscardWaitOrReady<Block, Self, Self::Blocked>, Error>;
 
 	/// called when the wait has completed. The canonical number is passed through
 	/// for further checks.
 	fn wait_completed(self, canon_number: NumberFor<Block>) -> Option<Self::Blocked>;
 }
 
+/// Describes whether a given [`BlockUntilImported`] (a) should be discarded, (b) is waiting for
+/// specific blocks to be imported or (c) is ready to be used.
+///
+/// A reason for discarding a [`BlockUntilImported`] would be if a referenced block is perceived
+/// under a different number than specified in the message.
+pub(crate) enum DiscardWaitOrReady<Block: BlockT, W, R> {
+	Discard,
+	Wait(Vec<(Block::Hash, NumberFor<Block>, W)>),
```

Comment:
A lot of the callback-based approach was to avoid this intermediate `Vec` but that is probably not a big deal